### PR TITLE
implement custom atexit() sequence

### PR DESCRIPTION
**Pull request by fabiangreffrath**

### CODE REVIEW

**Comment by fabiangreffrath** [comment]
I am seriously considering to remove the `run_on_error` parameter from `I_AtExit()`. First, we are currently using stdlib's `atexit()` which doesn't have any destinction between functions to run on regular exit and functions to run on error - and still things work as expected (or if they don't, it's not due to this). Second, I fail to imagine a faulty configuration key that would lead to a non-zero `exit()` call (segfaults don't count, configuration isn't saved here anyway) and that would let us want to leave the configuration unsaved.

File: Source/i_main.c [hunk 1]
```diff
@@ -87,7 +87,7 @@ void I_AtExit(atexit_func_t func, boolean run_on_error)
 {
     atexit_listentry_t *entry;
 
-    entry = malloc(sizeof(*entry));
+    entry = (malloc)(sizeof(*entry));
```
Comment (rfomin) [inline Source/i_main.c:90]: Not relevant to this PR, but how do you feel about removing memory function macros? I think they don't help and are actually very annoying.

**Review by rfomin — APPROVED** [review]
Works fine on Windows.

File: Source/i_main.c (hunk 1 above)
Comment (fabiangreffrath) [reply Source/i_main.c:90]: Yes, they often feel like shooting into the own foot. We would need to find out how often they are used implicitely in the code, though, i.e. by calling `malloc()` when indeed `Z_Malloc()` was meant.

**Comment by fabiangreffrath** [comment]
> I am seriously considering to remove the `run_on_error` parameter from `I_AtExit()`.

@rfomin @kraflab Opinions?

**Comment by rfomin** [comment]
> > I am seriously considering to remove the `run_on_error` parameter from `I_AtExit()`.
> 
> @rfomin @kraflab Opinions?

Sounds reasonable.

**Comment by kraflab** [comment]
The next release of dsda-doom is going to always attempt to write demo files on exit, even if an error occurred. I've actually already used this to create a demo that reproduces a freezing problem in a wad to help debug it. But that use case requires a signal handler, since I had to kill the process from outside. I'm not sure how much use it would be in other contexts.

File: Source/i_main.c (hunk 1 above)
Comment (kraflab) [reply Source/i_main.c:90]: These macros are the bane of my existence. There are places calling just malloc which is actually Z_Malloc but another file explicitly calls Z_Free on that data. Or vice versa, where something just calls free on another thing that was allocated with Z_Malloc. I had a branch adjusting all of this in dsda-doom but in the end I didn't want to risk having a mismatch in some context I wasn't aware of. It's probably cleaner in woof though.

**Comment by fabiangreffrath** [comment]
> I am seriously considering to remove the `run_on_error` parameter from `I_AtExit()`. 

I think I'll leave this in, but keep it at `true` for each call in the sequence (i.e. as it is now). Maybe one day we'll have a signal handler and have some better use for this. Then we don't have reimplement it again.

**Pull request merged**